### PR TITLE
ipPool, node deletion and affinity, blocklist and panic bugfix + tests

### PR DESCRIPTION
**Pull request by gunjan5**

### CODE REVIEW

File: lib/client/ipam_test.go [hunk 1]
```diff
@@ -103,6 +103,72 @@ type testArgsClaimAff struct {
...
+			v4, _, outErr := ic.AutoAssign(args)
+
+			It("assigned IP should be from pool1", func() {
+				Expect(outErr).NotTo(HaveOccurred())
+				Expect(pool1.IPNet.Contains(v4[0].IP)).To(BeTrue())
+			})
+		})
+
+		// Step-1.5: Have to create a node with name="host-A" so releasae affinity works.
+		c.Nodes().Create(&api.Node{
```
Comment (robbrockbank) [inline lib/client/ipam_test.go]: I don't understand why you need this.

File: lib/client/ipam_test.go [hunk 2]
```diff
@@ -103,6 +103,72 @@ type testArgsClaimAff struct {
...
+			Spec:     api.NodeSpec{},
+		})
+
+		// Step-2: Delete pool1 - expect it to execute without any error.
+		Context("Delete pool1", func() {
+			outErr := c.IPPools().Delete(api.IPPoolMetadata{
+				CIDR: pool1,
+			})
+
+			ic.ReleasePoolAffinities(pool1)
```
Comment (robbrockbank) [inline lib/client/ipam_test.go]: I'm not sure I follow what you are trying to test here.

Comment (gunjan5) [inline lib/client/ipam_test.go]: sorry shouldve removed this line (I thought I did)

File: lib/client/ipam_test.go [hunk 3]
```diff
@@ -103,6 +103,70 @@ type testArgsClaimAff struct {
...
+		// Step-2: Delete pool1 - expect it to execute without any error.
+		Context("Delete pool1", func() {
+			outErr := c.IPPools().Delete(api.IPPoolMetadata{
+				CIDR: pool1,
+			})
+
+			It("should delete the ipPool without any error", func() {
+				Expect(outErr).NotTo(HaveOccurred())
+			})
+		})
```
Comment (robbrockbank) [inline lib/client/ipam_test.go]: Is there anything we can do to check that it has removed affinities associated with those pools?

Comment (robbrockbank) [inline lib/client/ipam_test.go]: and also maybe assert that the affinity was correct before we removed the pool

File: lib/client/ipam_test.go [hunk 4]
```diff
@@ -103,6 +103,70 @@ type testArgsClaimAff struct {
...
+				Num6:     0,
+				Hostname: host,
+			}
+
+			v4, _, outErr := ic.AutoAssign(args)
+
+			It("assigned IP should be from pool1", func() {
+				Expect(outErr).NotTo(HaveOccurred())
+				Expect(pool1.IPNet.Contains(v4[0].IP)).To(BeTrue())
+			})
```
Comment (robbrockbank) [inline lib/client/ipam_test.go]: Add one more step to assert that another IP is assigned from the same /26

File: lib/client/ipam_test.go [hunk 5]
```diff
@@ -103,6 +103,70 @@ type testArgsClaimAff struct {
 
 var _ = Describe("IPAM tests", func() {
 
+	Describe("IPAM AutoAssign from the default pool then delete the pool and assign again", func() {
```
Comment (robbrockbank) [inline lib/client/ipam_test.go]: Maybe beef up comment to explain that we're testing you don't reuse affine block once pool is release.

**Comment by gunjan5** [comment]
🔙 to you @robbrockbank

**Comment by robbrockbank** [comment]
lgtm

**Comment by robbrockbank** [comment]
@gunjan5:  Can you squash the last four commits so that it's just the two commits?

**Comment by gunjan5** [comment]
Fixes #258 and some other bugs + proper node deletion behavior

**Pull request merged**